(chore): update developer documentation

Jipperism opened the pull request:
- Updates developer documentation for contributing to the sdk.
- Updates developer documentation for consuming the SDK.
- Removes redundant references to env vars.
- Removes redundant references to outdated package manager yarn.
- Updates typedoc version and configuration to include handwritten guides in the generated docs.
- Removes outdated github templates.
- Updates JSDoc annotations.

bitbeckers commented on .vscode/settings.json:
No .vscode settings in the repos please

Jipperism replied:
Removed them.

bitbeckers commented on CONTRIBUTING.md:
```diff
@@ -22,11 +22,11 @@ Before you submit a PR, make sure that:
 ### Release
 
 - Create a [personal access token](https://github.com/settings/tokens/new?scopes=repo&description=release-it) (Don't change the default scope)
 - Create an `.env` (copy `.env.template`) and set you github personal access token.
```
.env.example was removed, can't find .env.template?

bitbeckers commented on CONTRIBUTING.md:
```diff
@@ -22,11 +22,11 @@ Before you submit a PR, make sure that:
 
 ### Release
 
 - Create a [personal access token](https://github.com/settings/tokens/new?scopes=repo&description=release-it) (Don't change the default scope)
```
Are we using release-it?

Jipperism replied:
Removed the entire file as it's not up to date with our contribution guidelines.

bitbeckers commented on doc/_media/cancelOrders.md:
````diff
@@ -0,0 +1,69 @@
+```ts
+import { HypercertExchangeClient, ChainId } from "@hypercerts-org/marketplace-sdk";
+
+const hypercertExchangeClient = new HypercertExchangeClient(ChainId.MAINNET, provider, signer);
````
We're not on MAINNET.

Jipperism replied:
Changed it to OPTIMISM.

pheuberger reviewed (commented):
Looks like you're not only removing old code, but also bumping the version number and changing some magic number. I would advocate for even more granular commits, because I don't know why you're doing any of that (except the version bump).

pheuberger commented on src/HypercertExchangeClient.ts:
```diff
@@ -351,7 +319,7 @@ export class HypercertExchangeClient {
    * @param taker Taker order
    * @param signature Signature of the maker order
-   * @param merkleTree If the maker has been signed with a merkle tree
+   * @param merkleTree Optional merkle tree
```
Why would the user use a Merkle tree? I think that would be a much better parameter description.

Jipperism replied:
I had a more in-depth look, and I originally misinterpreted what the `merkleTree` was referring to. It seems to only refer to the proofs that are obtained when using `signMultipleMakerOrders()` to create the orders. See https://docs.looksrare.org/developers/sdk/signMultipleMakerOrders for a more thorough explanation.

I thought we would use this for allowlist orders, but they are implemented using a strategy and the merkle tree proofs for the allowlist are provided using the `additionalParams` argument. So it looks like we will only need this in the foreseeable future, if we plan on supporting signing multiple sales (and buys) at once from the frontend. That would also include implementing the endpoint for storing the tree and retrieving proofs (see https://docs.looksrare.org/developers/sdk/signMultipleMakerOrders again).

We could decide to remove the functionality, and that means we can get rid of all `merkleTree` parameters in `HypercertExchangeClient`. It would be a breaking change however, as the function signatures would change, but I think it might be worth it.

Looping in @bitbeckers and @holkexyz to weigh in on whether we are going to need this functionality in the future.

pheuberger replied:
Thank you for the thorough explanation! Also waiting for @bitbeckers and @holkexyz to chime in here.

bitbeckers replied:
I wouldn't invest time in these kinda of refactors. Is we're going to invest significant amounts of time in the marketplace SDK it would be migrate it to `viem` in the regular SDK.

bitbeckers replied:
Reading this again, we also have merkle proofs in strategies to allow for users to determin to whom they want to sell.

Jipperism replied:
The allowlist here refers to a different type. Namely, the one that gets created when using signMultipleMakerOrders from the SDK. This should only be used programatically according to the devs of looksrare.

So yes, in the future we might need a param merkleTree, but this is not referring to that specific one.

pheuberger commented on src/HypercertExchangeClient.ts:
```diff
@@ -766,7 +708,7 @@ export class HypercertExchangeClient {
 
   /**
-   * Register the order with hypercerts marketplace API.
+   * Register the order with hypercerts marketplace API
```
... with the Hypercerts ...

pheuberger commented on src/utils/api.ts:
```diff
@@ -124,9 +126,10 @@ export class ApiClient {
+  }
 
   /**
+   * @deprecated use GraphQL api instead instead
```
Same as above.

pheuberger commented on typedoc.json:
```diff
@@ -1,13 +1,12 @@
 {
   "$schema": "https://typedoc.org/schema.json",
-  "entryPoints": ["./src/HypercertExchangeClient.ts", "./src/types.ts"],
+  "entryPoints": ["./src/HypercertExchangeClient.ts", "./src/types.ts", "./guides/readme.md"],
```
Isn't it `README.md`?

pheuberger reviewed (commented):
Great stuff. Thank you so much for making this! It's mostly just nits and a couple of clarifying questions from my side. Being new to the project, I have big gaps in my knowledge and understanding which shows in my questions. Feel free to not elaborate on some of them if you think they're just too preposterous to cover hehe.

Please also don't be put off by the amount of comments. I was very thorough and highlighted very minor things. Generally this is great!

Jipperism commented on README.md:
```diff
@@ -1,101 +1,138 @@
+1. An order being executed or canceled (or many other reasons) might render it being invalid.
+    1. User B fetches order from API
+    1. User B creates taker order for maker order
+    1. User B signs taker order against maker order using the `HypercertExchangeClient`
```
Yes.

Jipperism commented on README.md:
```diff
@@ -1,101 +1,138 @@
+    1. User B creates taker order for maker order
+    1. User B signs taker order against maker order using the `HypercertExchangeClient`
+    1. User B calls the `executeOrder` method on the `HypercertExchangeClient` with the taker order, which calls the `HypercertExchange` contract in turn.
+    1. The `HypercertExchange` contract verifies the signature and executes the order
```
Yep, it is.

Jipperism replied:
Is that something that should be added?

pheuberger replied:
Not sure to be honest. I think it's fine the way it is 👍

Jipperism commented on README.md:
```diff
@@ -1,101 +1,138 @@
+
+> [!WARNING]
+>
+> The indexer is not listing to the OrderExecuted event yet, so the order will not be updated in the DB, unless reverified manually.
```
Correct.

Jipperism commented on guides/cancelOrders.md:
````diff
@@ -7,40 +14,16 @@
 This method is used to invalidate the `orderNonce`. If multiple maker orders share the same nonce, they will all be cancelled and non-executable. An `orderNonce` is also invalidated once an order with that nonce is executed.
 
 ```ts
-import { LooksRare, ChainId } from "@looksrare/sdk-v2";
````
Yep. We renamed the main export from `LooksRare` to `HypercertExchangeClient`. It's been like that since the start, but we didn't update the docs yet to reflect that.

pheuberger replied:
Got it, thank you!

Jipperism commented on guides/createDirectFractionsSaleMakerAsk.md:
```diff
@@ -0,0 +1,109 @@
+  startTime: Date.now() / 1000, // Use it to create an order that will be valid in the future (Optional, Default to now)
+  endTime: Date.now() / 1000 + 60 * 60, // If you use a timestamp in ms, the function will revert (this order will be valid for one hour)
+  itemIds: [fractionId], // ID of the fraction that will be put on sale
+  price: parseUnits("1", 18), // Price for the entire fraction, in this example we're selling the entire fraction for 1 USDC.
```
It refers to a fraction, with is a part of a hypercert. I'm not really sure how to rephrase this to be honest.

Jipperism replied:
Gave it a shot, let me know what you think.

pheuberger replied:
I think it's better.  What do you think @bitbeckers?

Jipperism commented on guides/createFractionalSaleMakerAsk.md:
```diff
@@ -0,0 +1,120 @@
+// The ID of the fraction that will be put on sale
+const fractionId = 13601086205829910384631083059047775411896320n;
+
+// Create the fractional sale order for a hypercert
```
Lowercase, see https://www.hypercerts.org/docs/what-are-hypercerts

Jipperism commented on guides/createFractionalSaleMakerAsk.md:
```diff
@@ -0,0 +1,120 @@
+const order = await hypercertExchangeClient.createFractionalSaleMakerAsk({
+  startTime: Date.now() / 1000, // Use it to create an order that will be valid in the future (Optional, Default to now)
+  endTime: Date.now() / 1000 + 60 * 60, // If you use a timestamp in ms, the function will revert (this order will be valid for one hour)
+  price: '10', // Note that this is the price per hypercert unit in wei
```
Went through the docs, the closest I can find is something like https://www.hypercerts.org/docs/guide/sell#advanced-settings. Let me know what you think. Looping in @bitbeckers in case I missed it somewhere in the docs.

pheuberger replied:
It's probably better than not having it. Perhaps it's not worth explaining it in more detail. I think what trips me up is that there are fractions of a hypercert and then you can also divide it up even further by selling/buying fractions (a number of units) out of that fraction.

bitbeckers replied:
There's also documentation on this on the specific strategy contract. It might be good to explain here what the strategy does

Jipperism replied:
Addressed

Jipperism commented:
@pheuberger Had a look at all the comments and implemented the feedback. Please close all conversations on the PR that you consider resolved.

> Looks like you're not only removing old code, but also bumping the version number and changing some magic number. I would advocate for even more granular commits, because I don't know why you're doing any of that (except the version bump).

I updated the git commit description for https://github.com/hypercerts-org/marketplace-sdk/pull/25/commits/fe6ed4b8b680518fc64c081ed87f8ae3f39f92c4, hopefully that makes things clearer. I think splitting this up in more granular commits would also entail purposely releasing broken versions to NPM, as each commit would require a version bump (otherwise we have version on github that don't exist on NPM)? I can see both sides of the story but maybe that's not necessary for now, let me know what you think.

pheuberger commented on README.md:
```diff
@@ -1,101 +1,137 @@
+
+### Lifecycle of an order
+
+1. User A creates maker order and signs it
```
All the numbered bullets are the same (1.)

Jipperism replied:
Yep, but markdown renders that as a numbered list correctly. This way it's easier to add something in the middle if necessary.

Example: (I think you can click edit on this comment to see the source).
1. a
1. b
1. c

pheuberger replied:
Ah, that's great, didn't know!

pheuberger commented on typedoc.json:
```diff
@@ -1,13 +1,12 @@
 
-  "hideBreadcrumbs": true,
-  "hideInPageTOC": true
+  "projectDocuments": ["./guides/readme.md"],
```
This one should also be all caps like above, shouldn't it?

Jipperism replied:
Addressed

pheuberger commented on README.md:
```diff
@@ -1,101 +1,138 @@
+    1. User B signs taker order against maker order using the `HypercertExchangeClient`
+    1. User B calls the `executeOrder` method on the `HypercertExchangeClient` with the taker order, which calls the `HypercertExchange` contract in turn.
+    1. The `HypercertExchange` contract verifies the signature and executes the order
+    1. The `HypercertExchange` contract will emit an event that the order has been executed, which is picked up by the indexer which revalidates the order and updates the errors codes and validity in the DB.
```
It looks like there's something missing in the changed sentence @Jipperism, or am I missing something?

`If the transaction rendered the order invalid for any further sales, the error codes and validity in the DB.`

Jipperism replied:
For example: fraction doesn't have enough units left in it to do another sale. Also, fixed the missing part in the sentence.

pheuberger commented on guides/faq.md:
```diff
@@ -0,0 +1,24 @@
+
+### ❓ Why do I need to call grantTransferManagerApproval ?
+
+When you approve a collection to be traded on the hypercerts exchange, you approve the TransferManager instead of the exchange. Calling `grantTransferManagerApproval` gives the exchange contract the right to call the transfer function on the TransferManager. You need to call this function only once, the first time you use the V2.
```
It looks like you removed V2. Was that a leftover from LR?

Jipperism replied:
Yes, it was.

pheuberger commented:
> @pheuberger Had a look at all the comments and implemented the feedback. Please close all conversations on the PR that you consider resolved.
> 
> > Looks like you're not only removing old code, but also bumping the version number and changing some magic number. I would advocate for even more granular commits, because I don't know why you're doing any of that (except the version bump).
> 
> I updated the git commit description for [fe6ed4b](https://github.com/hypercerts-org/marketplace-sdk/commit/fe6ed4b8b680518fc64c081ed87f8ae3f39f92c4), hopefully that makes things clearer. I think splitting this up in more granular commits would also entail purposely releasing broken versions to NPM, as each commit would require a version bump (otherwise we have version on github that don't exist on NPM)? I can see both sides of the story but maybe that's not necessary for now, let me know what you think.

Thank you, @Jipperism! Can you explain why we would need to release broken versions to NPM? Not every commit needs to be a release, nor have a version bump, unless I'm missing something?

pheuberger commented on README.md:
```diff
@@ -1,101 +1,137 @@
+cd marketplace-sdk
+pnpm install && pnpm run build
+cd ../your-project
+pnpm link @hypercerts-org/marketplace-sdk
```
This is actually wrong. It needs to say 

```bash
pnpm link <path to marketplace sdk on you local filesystem>
```

Jipperism replied:
Addressed

bitbeckers commented on LICENSE.md:
License should also acknowledge Looksrare source repo

Jipperism replied:
Addressed

bitbeckers commented on README.md:
````diff
@@ -1,101 +1,137 @@
-yarn add @looksrare/sdk-v2 ethers
-```
+<!-- TODO: Check if links still work on deployment -->
+Read the [guides](./guides) if you need help with the implementation.
````
These will reference to the hypercerts.org/docs, right?

Jipperism replied:
Yes, will update this when moving the guides to @hypercerts-org/docs.

bitbeckers commented on README.md:
```diff
@@ -1,101 +1,137 @@
+- `build` - Build the SDK
+- `test` - Run the tests
+- `docs` - Generate the documentation into the `doc` folder
+- `supabase:types:hypercerts` - Generate types for the `data-staging` database
```
Due to my late review these are now outdated? Because we connect to the API, not Supabase?

Jipperism replied:
Yes and no. We still use the generated types in 1 or 2 places, but never connect to supabase anymore once https://github.com/hypercerts-org/marketplace-sdk/pull/27 gets merged. It could be removed if we pull out the 1 or 2 types that we're still using but I think that belongs in the other PR.

bitbeckers replied:
https://github.com/hypercerts-org/marketplace-sdk/issues/29

bitbeckers commented on README.md:
```diff
@@ -1,101 +1,137 @@
+### Data
 
-🔗 [Developer discord](https://discord.gg/jJA4qM5dXz)
+Order information and nonces live in the `data-staging` and `data-production` database. Nonces are invalidated on-chain, but keeping track of the current nonce for a user happens off-chain in the DB. Orders do not live on-chain, but as they are signed they can be verified.
```
Don't think this is relevant for a readme. If somebody forks this and wants to use a differen DB setup that's fine

Jipperism replied:
Addressed

bitbeckers commented on guides/cancelOrders.md:
```diff
@@ -49,25 +32,21 @@ const receipt = await tx.wait();
-## Need help?
-
-You can reach out to the LooksRare team via our Developers Discord: [https://discord.gg/LooksRareDevelopers](https://discord.gg/LooksRareDevelopers)
+For more details on the triple-nonce system, see the related documentation: [Triple-nonce system](https://docs.looksrare.org/developers/protocol/triple-nonce-system-v2)
```
Let's host that page ourselves

bitbeckers replied:
Write our own? What if their page goes down?

Jipperism replied:
Added the docs.

Jipperism replied:
@bitbeckers added it back after discussion with @holkexyz . Kept the missing subset nonce docs in there as well commented out, so we can show that once we've implemented the API part of it.

bitbeckers commented on guides/executeTrade.md:
```diff
@@ -1,26 +1,58 @@
 # How to create a Taker order and execute a trade
 
-Trades are executed on-chain by matching a `Maker` order with a `Taker` order. The maker order can be retrieved from the API, see [get v2 orders](https://looksrare.dev/v2/reference/getorders) for more details. While the taker order can be obtained by calling the `createTaker` method as shown here:
+Trades are executed on-chain by matching a `Maker` order with a `Taker` order. The maker order can be retrieved from the [graphQL API](https://staging-api.hypercerts.org/v1/graphql). The order used here has the same structure as a [maker order](../src/types.ts#L116), but with an added field `signature: string`. While the taker order can be obtained by calling the `createTaker` method as shown her ...
```
Link to production API, or both. But staging could always be broken and not used as reference

Jipperism replied:
Addressed

bitbeckers commented on src/HypercertExchangeClient.ts:
```diff
@@ -209,7 +201,7 @@ export class HypercertExchangeClient {
       price: price,
       itemIds: itemIds,
-      amounts: amounts,
+      amounts: itemIds.map(_ => 1n),
```
Does this make sense? is it always 1? Will this break when we enable other strategies?

bitbeckers commented on src/utils/api.ts:
```diff
@@ -4,6 +4,7 @@ import { Database as HypercertsDatabase } from "./hypercerts-database-types";
 import { CONSTANTS, parseClaimOrFractionId } from "@hypercerts-org/sdk";
 import { getFractionsById, getOrders } from "./graphl";
 import { cacheExchange, Client, fetchExchange } from "@urql/core";
+import { deprecate } from "util";
```
unused import? Type from @deprecated?

Jipperism replied:
Good catch. Got imported automatically when trying to mark a method as deprecated in the js docs.

Jipperism commented:
@bitbeckers GitHub UI is somehow not letting me respond to

>Reading this again, we also have merkle proofs in strategies to allow for users to determin to whom they want to sell.

The allowlist here refers to a different type. Namely, the one that gets created when using `signMultipleMakerOrders` from the SDK. This should only be used programatically according to the devs of looksrare.

So yes, in the future we might need a param `merkleTree`, but this is not referring to that specific one.

bitbeckers commented on src/HypercertExchangeClient.ts:
```diff
@@ -209,7 +201,7 @@ export class HypercertExchangeClient {
       price: price,
       itemIds: itemIds,
-      amounts: amounts,
+      amounts: itemIds.map(_ => 1n),
```
Resolved the conversation, but this is an update in business logic. So create a unit test, release an alpha and validate on testnet.

Jipperism replied:
Tested and works in https://github.com/hypercerts-org/hypercerts-app/pull/428.

bitbeckers replied:
Wait they're completely broken? Let's create an issue for that, we should be able to get green tests and automate package updates from our CI

Jipperism replied:
Yes, they have been almost since the beginning, and we've spent some time trying to get it to work as well but that was forever ago. Biggest issue is hardhat setup within the tests.

Jipperism replied:
Created an issue here: https://github.com/hypercerts-org/marketplace-sdk/issues/30

bitbeckers commented on src/utils/calls/nonces.ts:
```diff
@@ -30,23 +30,6 @@ export const cancelOrderNonces = (
   };
 };
 
-export const cancelSubsetNonces = (
```
Missed this last time, why was it removed?

Jipperism replied:
Because we don't support subset nonces in our api.

bitbeckers commented on guides/faq.md:
```diff
@@ -0,0 +1,24 @@
+
+### ❓ How to retrieve order nonce ?
+
+Call the public API endpoint [/marketplace/order-nonce](https://api.hypercerts.org/spec/#/Marketplace/UpdateOrderNonce), and use this nonce directly.
```
/v1/

Jipperism replied:
Fixed

bitbeckers commented on guides/faq.md:
```diff
@@ -0,0 +1,24 @@
+
+### ❓ What to do when the order is created and signed ?
+
+Use the public API endpoint [/marketplace/orders](https://api.hypercerts.org/spec/#/Marketplace/StoreOrder) to push the order to the database. After that, the order will be visible by everyone using the API. There is also a `hypercertExchangeClient.registerOrder()` utility method available.
```
/v1/

Jipperism replied:
Fixed

bitbeckers commented on guides/faq.md:
```diff
@@ -0,0 +1,24 @@
+
+🔗 [Public API documentation](https://api.hypercerts.org/spec)
+
+🔗 [Developer discord](https://discord.gg/PZtJ6W9S)
```
Invite invalid

Jipperism replied:
Fixed, created a non-expiring invite.

bitbeckers commented on guides/createMakerAsk.md:
```diff
@@ -1,57 +1,71 @@
-  collection: "0x0000000000000000000000000000000000000000", // Collection address
-  collectionType: CollectionType.ERC721,
+const { maker, isCollectionApproved, isTransferManagerApproved } = await hypercertExchangeClient.createMakerAsk({
+  collection: "0x123", // Collection address
```
Same pet peeve as in other docs, actually use addresses because everytime I read this I think the example is incorrect

bitbeckers commented on guides/createFractionalSaleMakerAsk.md:
```diff
@@ -0,0 +1,122 @@
+// Create the fractional sale order for a hypercert
+// This will also take care of fetching and setting the order nonce correctly
+const order = await hypercertExchangeClient.createFractionalSaleMakerAsk({
+  startTime: Date.now() / 1000, // Use it to create an order that will be valid in the future (Optional, Default to now)
```
We discussed yesterday that that would fail right? Taking blocktimes and on chain validations into account?

Jipperism replied:
Fixed, made it 5 minutes into the future now.

bitbeckers commented on guides/createDirectFractionsSaleMakerAsk.md:
```diff
@@ -0,0 +1,111 @@
+  startTime: Date.now() / 1000, // Use it to create an order that will be valid in the future (Optional, defaults to the current timestamp in seconds)
+  endTime: Date.now() / 1000 + 60 * 60, // If you use a timestamp in ms, the function will revert (this order will be valid for one hour)
+  itemIds: [fractionId], // ID of the fraction that will be put on sale
+  price: parseUnits("1", 18), // Price for all units in the hypercert fraction, in this example we're selling for 1 USDC.
```
actually wrong decimals for USDC

Jipperism replied:
Fixed. Impressed with the catch.

Jipperism commented on guides/createFractionalSaleMakerAsk.md:
```diff
@@ -0,0 +1,120 @@
+  await tx.wait();
+}
+
+// Approve the collection items to be transferred by the TransferManager
```
A collection in this context is not the same as our own collections. In the case of hypercerts, they all belong to the collection [0xa16DFb32Eb140a6f3F2AC68f41dAd8c7e83C4941](https://sepolia.etherscan.io/address/0xa16DFb32Eb140a6f3F2AC68f41dAd8c7e83C4941) (hypercert-minter-address). The original usecase, it would probably refer to the NFTs minter address.

bitbeckers commented:
As you've release 0.4.0 that'd also need to be merged in here since this PR also influences business logic

Jipperism commented:
@bitbeckers has been updated to incorporate changes from 0.4.1
